sweep/sweeper.go

```diff
@@ -497,211 +466,104 @@ func (s *UtxoSweeper) signalAndRemove(outpoint *wire.OutPoint, result Result) {
-				outputValue, s.dustLimit)
-			return false, nil
-		}
+	var err error
 
-		// If fullOnly, don't sweep if less inputs than the maximum. Do
-		// this check after the dust limit check, to properly report
-		// whether there are sweepable inputs remaining.
-		if fullOnly && len(inputList) < s.cfg.MaxInputsPerTx {
-			return true, nil
+	// Generate output script if no unused script available.
+	if s.currentOutputScript == nil {
+		s.currentOutputScript, err = s.cfg.GenSweepScript()
+		if err != nil {
+			return err
 		}
+	}
 
-		s.sweep(inputList, uint32(currentHeight), satPerKW)
+	// Create sweep tx.
+	tx, err := s.cfg.TxGenerator.createSweepTx(inputs,
+		s.currentOutputScript,
+		uint32(currentHeight), satPerKW)
+	if err != nil {
+		return fmt.Errorf("cannot create sweep tx: %v", err)
 	}
 
-	return false, nil
-}
+	// Add tx before publication, so that we will always know that a
+	// spend by this tx is ours.
+	s.cfg.Store.NotifyPublishTx(tx)
 
-// sweep creates and publishes a sweep tx for the provided list of inputs. It
-// uses the outcome to advance the pending input states and notifies listeners
-// if necessary.
-func (s *UtxoSweeper) sweep(inputList []Input, currentHeight uint32,
-	feePerKw lnwallet.SatPerKWeight) {
+	// Publish sweep tx.
+	log.Debugf("Publishing sweep tx %v", tx.TxHash())
+	err = s.cfg.PublishTransaction(tx)
 
-	sweepTx, err := s.createAndPublishTx(inputList, currentHeight, feePerKw)
+	// Keep outputScript in case of an error, so that it can be
+	// reused for the next tx and causes no address inflation.
 	if err != nil {
 		log.Error(err)
+	} else {
+		s.currentOutputScript = nil
 	}
 
-	// TODO: Rebroadcaster/retry mechanisms need to be added.
-
-	// Advance input states or report outcomes.
-	for _, input := range inputList {
-		pi := s.pendingInputs[*(input.OutPoint())]
-
-		switch {
+	// Reschedule sweep.
+	for _, input := range tx.TxIn {
+		pi := s.pendingInputs[input.PreviousOutPoint]
 
-		// Published successfully, record tx hash.
-		case err == nil:
-			pi.txHash = sweepTx.TxHash()
-			pi.state = statePublished
+		var nextAttemptDelta int32
+		switch err {
+		// Successful publish, but it may still not confirm.
+		// Retry later.
+		case nil:
+			nextAttemptDelta = 6
 
-		// In case of a double spend, don't notify listeners yet
-		// but wait for the spending tx to trigger the spend
-		// notification. Because of the inaccurate error
-		// reporting, ErrDoubleSpend may actually mean various
-		// things other than a real double spend.
-		case err == lnwallet.ErrDoubleSpend:
-			pi.state = stateError
+		// In case of a double spend, retry the next block.
+		case lnwallet.ErrDoubleSpend:
+			nextAttemptDelta = 1
 
-		// In case of an other error, signal the listener and
-		// remove the input.
+		// For other errors, back off exponentially.
 		default:
-			s.signalAndRemove(input.OutPoint(), Result{Err: err})
+			nextAttemptDelta = 1 << uint(pi.publishAttempts)
```

Comment:
Cap the back off at a point in the future?

Reply:
Also at one point to we give up? Never?

Reply:
Gotcha, I think we should still cap, and then print the error each attempt it fails. That would be better than now where if a failure to sweep occurs, then it's quickly buried in logs to likely never be seen again after a log rotation.

Reply:
Cap added, currently at 144 attempts (=blocks)